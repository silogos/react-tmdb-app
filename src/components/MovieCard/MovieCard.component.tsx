import { NavLink } from "react-router";
import { getReleaseYear } from "@/utils/Movie.utils";

import { MovieCardProps } from "./MovieCard.type";
import Poster from "@/components/Poster";
import { StarIcon } from "@heroicons/react/24/solid";

const _renderInfo = ({ title, releaseDate }: MovieCardProps) => {
  return (
    <>
      <div className="text-base mb-1">{title}</div>
      <div className="text-sm text-gray-400">{getReleaseYear(releaseDate)}</div>
    </>
  );
};

function MovieCard(props: MovieCardProps) {
  const { id, poster, voteAverage } = props;

  return (
    <NavLink
      to={`/movie/${id}`}
      className="flex flex-col gap-3 group relative overflow-hidden cursor-pointer rounded-lg"
    >
      <Poster poster={poster}>
        <div className="hidden lg:flex duration-150 group-hover:opacity-100 opacity-0 absolute inset-0 justify-center items-center z-20 bg-black bg-opacity-50" />
        <div className="hidden lg:flex absolute bottom-0 group-hover:opacity-100 opacity-0 duration-150 w-full flex-col p-4 text-white bg-black bg-opacity-50 z-30">
          {_renderInfo(props)}
        </div>
        <div className="lg:group-hover:hidden absolute top-3 right-3 px-2 py-1 bg-gray-50 rounded-lg flex flex-col self-start justify-center items-center text-center text-xs text-gray-900 font-medium mb-1">
          <StarIcon className="size-5" />
          {voteAverage.toFixed(1)}
        </div>
      </Poster>
      <div className="flex lg:hidden flex-col text-gray-300 w-auto">
        {_renderInfo(props)}
      </div>
    </NavLink>
  );
}

export default MovieCard;
