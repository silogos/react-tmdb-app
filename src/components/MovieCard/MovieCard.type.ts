export type MovieCardProps = {
  id: number;
  title: string;
  releaseDate: string;
  poster: string;
};
