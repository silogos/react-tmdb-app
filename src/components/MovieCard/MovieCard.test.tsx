import { describe, it, expect, vi } from "vitest";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router";
import MovieCard from "./MovieCard.component";
import { MovieCardProps } from "./MovieCard.type";

vi.mock("@/components/Poster");

describe("MovieCard Component", () => {
  const movie: MovieCardProps = {
    id: 1,
    title: "Inception",
    releaseDate: "2010-07-16",
    poster: "inception.jpg",
    voteAverage: 1,
  };

  it("renders correctly", () => {
    const { container, getAllByText } = render(
      <BrowserRouter>
        <MovieCard {...movie} />
      </BrowserRouter>
    );

    expect(getAllByText(/Inception/)).toHaveLength(2);
    expect(getAllByText("2010")).toHaveLength(2);
    expect(container).toMatchSnapshot();
  });

  it("has correct link", () => {
    const { container } = render(
      <BrowserRouter>
        <MovieCard {...movie} />
      </BrowserRouter>
    );

    const link = container.querySelector("a");
    expect(link).toHaveAttribute("href", "/movie/1");
  });
});
