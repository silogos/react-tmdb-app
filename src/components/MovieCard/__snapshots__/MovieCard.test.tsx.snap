// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MovieCard Component > renders correctly 1`] = `
<div>
  <a
    class="flex flex-col gap-3 group relative overflow-hidden cursor-pointer rounded-lg"
    data-discover="true"
    href="/movie/1"
  >
    <div
      data-testid="Poster"
    >
      <div
        class="hidden lg:flex duration-150 group-hover:opacity-100 opacity-0 absolute inset-0 justify-center items-center z-20 bg-black bg-opacity-50"
      />
      <div
        class="hidden lg:flex absolute bottom-0 group-hover:opacity-100 opacity-0 duration-150 w-full flex-col p-4 text-white bg-black bg-opacity-50 z-30"
      >
        <div
          class="text-base mb-1"
        >
          Inception
        </div>
        <div
          class="text-sm text-gray-400"
        >
          2010
        </div>
      </div>
      <div
        class="lg:group-hover:hidden absolute top-3 right-3 px-2 py-1 bg-gray-50 rounded-lg flex flex-col self-start justify-center items-center text-center text-xs text-gray-900 font-medium mb-1"
      >
        <svg
          aria-hidden="true"
          class="size-5"
          data-slot="icon"
          fill="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M10.788 3.21c.448-1.077 1.976-1.077 2.424 0l2.082 5.006 5.404.434c1.164.093 1.636 1.545.749 2.305l-4.117 3.527 1.257 5.273c.271 1.136-.964 2.033-1.96 1.425L12 18.354 7.373 21.18c-.996.608-2.231-.29-1.96-1.425l1.257-5.273-4.117-3.527c-.887-.76-.415-2.212.749-2.305l5.404-.434 2.082-5.005Z"
            fill-rule="evenodd"
          />
        </svg>
        1.0
      </div>
    </div>
    <div
      class="flex lg:hidden flex-col text-gray-300 w-auto"
    >
      <div
        class="text-base mb-1"
      >
        Inception
      </div>
      <div
        class="text-sm text-gray-400"
      >
        2010
      </div>
    </div>
  </a>
</div>
`;
